File: src/lib/recommendations/rules.ts
import { db } from "@/db";
import { hives, inspections, feedings, equipmentDeployments, equipmentStock, equipmentTypes } from "@/db/schema";
import { eq, isNull, desc, sql } from "drizzle-orm";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type RecommendationType =
  | "inspection_due"
  | "treatment_reminder"
  | "equipment_needed"
  | "seasonal_prep"
  | "feeder_check";

export interface RuleResult {
  hiveId: string;
  message: string;
  priority: "low" | "normal" | "high" | "urgent";
}

export interface RuleContext {
  /** Current date (injectable for testing) */
  now: Date;
}

export interface RecommendationRule {
  type: RecommendationType;
  check: (context: RuleContext) => Promise<RuleResult[]>;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function daysBetween(a: Date, b: Date): number {
  return Math.floor((b.getTime() - a.getTime()) / (1000 * 60 * 60 * 24));
}

/**
 * Estimate treatment duration in days based on method/type string.
 * Conservative defaults so we remind rather than miss.
 */
function estimateTreatmentDurationDays(method: string): number {
  const lower = (method ?? "").toLowerCase();
  if (lower.includes("oxalic") && lower.includes("vaporiz")) return 7;
  if (lower.includes("oxalic")) return 1; // oxalic acid dribble — single application
  if (lower.includes("apivar") || lower.includes("amitraz")) return 42;
  if (lower.includes("apiguard") || lower.includes("thymol")) return 28;
  if (lower.includes("formic") || lower.includes("mite away")) return 14;
  if (lower.includes("hopguard")) return 30;
  if (lower.includes("checkmite") || lower.includes("coumaphos")) return 42;
  // Default: 14 days
  return 14;
}

// ---------------------------------------------------------------------------
// Rule: inspection_due
// ---------------------------------------------------------------------------

const inspectionDueRule: RecommendationRule = {
  type: "inspection_due",
  async check({ now }) {
    const DEFAULT_INTERVAL_DAYS = 14;
    const results: RuleResult[] = [];

    // Get all active hives
    const activeHives = await db
      .select({ id: hives.id, positionLabel: hives.positionLabel })
      .from(hives)
      .where(eq(hives.status, "active"));

    for (const hive of activeHives) {
      // Get most recent inspection for this hive
      const latest = await db
        .select({ date: inspections.date })
        .from(inspections)
        .where(eq(inspections.hiveId, hive.id))
        .orderBy(desc(inspections.date))
        .limit(1);

      const lastDate = latest[0]?.date;
      if (!lastDate) {
        // Never inspected
        results.push({
          hiveId: hive.id,
          message: `Hive "${hive.positionLabel}" has never been inspected. Schedule an inspection soon.`,
          priority: "high",
        });
        continue;
      }

      const overdueDays = daysBetween(lastDate, now) - DEFAULT_INTERVAL_DAYS;

      if (overdueDays > 14) {
        results.push({
          hiveId: hive.id,
          message: `Hive "${hive.positionLabel}" is ${overdueDays + DEFAULT_INTERVAL_DAYS} days since last inspection. Immediate inspection recommended.`,
          priority: "urgent",
        });
      } else if (overdueDays > 7) {
        results.push({
          hiveId: hive.id,
          message: `Hive "${hive.positionLabel}" is ${overdueDays + DEFAULT_INTERVAL_DAYS} days since last inspection.`,
          priority: "high",
        });
      } else if (overdueDays > 0) {
        results.push({
          hiveId: hive.id,
          message: `Hive "${hive.positionLabel}" inspection is overdue by ${overdueDays} days.`,
          priority: "normal",
        });
      }
    }

    return results;
  },
};

// ---------------------------------------------------------------------------
// Rule: treatment_reminder
// ---------------------------------------------------------------------------

const treatmentReminderRule: RecommendationRule = {
  type: "treatment_reminder",
  async check({ now }) {
    const results: RuleResult[] = [];

    // Find inspections with treatments recorded
    const rows = await db
      .select({
        id: inspections.id,
        hiveId: inspections.hiveId,
        date: inspections.date,
        treatments: inspections.treatments,
        hiveName: hives.positionLabel,
      })
      .from(inspections)
      .innerJoin(hives, eq(inspections.hiveId, hives.id))
      .where(eq(hives.status, "active"));

    // Group by hive — only keep the latest inspection with treatments
    const hiveLatest = new Map<
      string,
      { date: Date; treatments: unknown; hiveName: string }
    >();

    for (const row of rows) {
      if (!row.treatments) continue;

      const existing = hiveLatest.get(row.hiveId);
      if (!existing || row.date > existing.date) {
        hiveLatest.set(row.hiveId, {
          date: row.date,
          treatments: row.treatments,
          hiveName: row.hiveName,
        });
      }
    }

    for (const [hiveId, data] of hiveLatest) {
      // treatments is JSONB — could be array of objects or other shapes
      const treatmentList = Array.isArray(data.treatments)
        ? data.treatments
        : [data.treatments];

      for (const t of treatmentList) {
        const method =
          typeof t === "string"
            ? t
            : typeof t === "object" && t !== null
              ? ((t as Record<string, unknown>).method as string) ??
                ((t as Record<string, unknown>).name as string) ??
                ((t as Record<string, unknown>).type as string) ??
                JSON.stringify(t)
              : String(t);

        const durationDays = estimateTreatmentDurationDays(method);
        const daysSinceTreatment = daysBetween(data.date, now);

        if (daysSinceTreatment >= durationDays) {
          const overdue = daysSinceTreatment - durationDays;
          results.push({
            hiveId,
            message: `Treatment "${method}" on hive "${data.hiveName}" may need follow-up. Applied ${daysSinceTreatment} days ago (estimated ${durationDays}-day duration).`,
            priority: overdue > 7 ? "high" : "normal",
          });
        }
      }
    }

    return results;
  },
};

// ---------------------------------------------------------------------------
// Rule: equipment_needed
// ---------------------------------------------------------------------------

const equipmentNeededRule: RecommendationRule = {
  type: "equipment_needed",
  async check() {
    const results: RuleResult[] = [];

    // Frame shortage per hive from active deployments: box deployments
    // provide capacity (framesPerBox), frame deployments fill it.
    const deployed = await db
      .select({
        hiveId: equipmentDeployments.hiveId,
        hiveName: hives.positionLabel,
        quantity: equipmentDeployments.quantity,
        category: equipmentTypes.category,
        framesPerBox: equipmentTypes.framesPerBox,
      })
      .from(equipmentDeployments)
      .innerJoin(equipmentStock, eq(equipmentDeployments.stockId, equipmentStock.id))
      .innerJoin(equipmentTypes, eq(equipmentStock.typeId, equipmentTypes.id))
      .innerJoin(hives, eq(equipmentDeployments.hiveId, hives.id))
      .where(
        sql`${equipmentDeployments.dateRemoved} is null and ${hives.status} = 'active'`
      );

    // Aggregate per hive
    const hiveShortages = new Map<
      string,
      { hiveName: string; totalCapacity: number; totalInstalled: number }
    >();

    for (const row of deployed) {
      const entry = hiveShortages.get(row.hiveId) ?? {
        hiveName: row.hiveName,
        totalCapacity: 0,
        totalInstalled: 0,
      };
      if (row.category === "box" && row.framesPerBox) {
        entry.totalCapacity += row.quantity * row.framesPerBox;
      } else if (row.category === "frame") {
        entry.totalInstalled += row.quantity;
      }
      hiveShortages.set(row.hiveId, entry);
    }

    for (const [hiveId, data] of hiveShortages) {
      const shortage = data.totalCapacity - data.totalInstalled;
      if (shortage > 0) {
        const pct = Math.round((shortage / data.totalCapacity) * 100);
        results.push({
          hiveId,
          message: `Hive "${data.hiveName}" needs ${shortage} more frames (${pct}% capacity unfilled).`,
          priority: pct > 50 ? "high" : "normal",
        });
      }
    }

    return results;
  },
};

// ---------------------------------------------------------------------------
// Rule: feeder_check
// ---------------------------------------------------------------------------

const feederCheckRule: RecommendationRule = {
  type: "feeder_check",
  async check({ now }) {
    const DEFAULT_CHECK_DAYS = 7;
    const results: RuleResult[] = [];

    // Active feeders: dateEmpty is null
    const activeFeedings = await db
      .select({
        id: feedings.id,
        hiveId: feedings.hiveId,
        dateFed: feedings.dateFed,
        type: feedings.type,
        hiveName: hives.positionLabel,
      })
      .from(feedings)
      .innerJoin(hives, eq(feedings.hiveId, hives.id))
      .where(isNull(feedings.dateEmpty));

    for (const f of activeFeedings) {
      const daysSinceFed = daysBetween(f.dateFed, now);
      if (daysSinceFed >= DEFAULT_CHECK_DAYS) {
        results.push({
          hiveId: f.hiveId,
          message: `Feeder on hive "${f.hiveName}" (${f.type.replace(/_/g, " ")}) placed ${daysSinceFed} days ago. Check if empty or needs refill.`,
          priority: daysSinceFed > 14 ? "high" : "normal",
        });
      }
    }

    return results;
  },
};

// ---------------------------------------------------------------------------
// Rule: seasonal_prep
// ---------------------------------------------------------------------------

const seasonalPrepRule: RecommendationRule = {
  type: "seasonal_prep",
  async check({ now }) {
    const results: RuleResult[] = [];
    const month = now.getMonth(); // 0-indexed

    // Get all active hives for seasonal recommendations
    const activeHives = await db
      .select({ id: hives.id, positionLabel: hives.positionLabel })
      .from(hives)
      .where(eq(hives.status, "active"));

    if (activeHives.length === 0) return results;

    let seasonalMessage: string | null = null;
    let priority: RuleResult["priority"] = "normal";

    // Spring buildup: Feb (1) - Mar (2)
    if (month === 1 || month === 2) {
      seasonalMessage =
        "Spring buildup season. Check food stores, assess colony strength, and consider supplemental feeding if stores are low.";
      priority = "normal";
    }
    // Swarm season: Apr (3) - May (4)
    else if (month === 3 || month === 4) {
      seasonalMessage =
        "Swarm season alert. Inspect for swarm cells, ensure adequate space, and consider splits for strong colonies.";
      priority = "high";
    }
    // Fall prep: Sep (8) - Oct (9)
    else if (month === 8 || month === 9) {
      seasonalMessage =
        "Fall preparation time. Assess winter stores, treat for varroa if needed, and reduce entrances.";
      priority = "high";
    }
    // Winter check: Dec (11) - Jan (0)
    else if (month === 11 || month === 0) {
      seasonalMessage =
        "Winter monitoring period. Check hive weight for adequate stores, ensure ventilation, and minimize disturbance.";
      priority = "low";
    }

    if (seasonalMessage) {
      // Create one recommendation per hive
      for (const hive of activeHives) {
        results.push({
          hiveId: hive.id,
          message: `${seasonalMessage} (Hive: ${hive.positionLabel})`,
          priority,
        });
      }
    }

    return results;
  },
};

// ---------------------------------------------------------------------------
// Export all rules
// ---------------------------------------------------------------------------

export const recommendationRules: RecommendationRule[] = [
  inspectionDueRule,
  treatmentReminderRule,
  equipmentNeededRule,
  feederCheckRule,
  seasonalPrepRule,
];

export { daysBetween, estimateTreatmentDurationDays };

File: backend/internal/recs/rules.go
// Package recs implements the recommendations rules engine ported from the
// legacy TypeScript implementation (src/lib/recommendations/{rules,engine}.ts).
package recs

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Result is a single recommendation produced by a rule.
type Result struct {
	// HiveID is the target hive (nil = apiary-wide).
	HiveID   *string
	Message  string
	Priority string // low | normal | high | urgent
}

type rule struct {
	Type  string
	Check func(ctx context.Context, pool *pgxpool.Pool, now time.Time) ([]Result, error)
}

func allRules() []rule {
	return []rule{
		{Type: "inspection_due", Check: checkInspectionDue},
		{Type: "treatment_reminder", Check: checkTreatmentReminder},
		{Type: "equipment_needed", Check: checkEquipmentNeeded},
		{Type: "feeder_check", Check: checkFeederCheck},
		{Type: "seasonal_prep", Check: checkSeasonalPrep},
	}
}

// RuleCount reports how many rules the engine evaluates per run.
func RuleCount() int { return len(allRules()) }

// daysBetween returns the floor of the day difference b - a.
func daysBetween(a, b time.Time) int {
	return int(math.Floor(b.Sub(a).Hours() / 24))
}

// estimateTreatmentDurationDays estimates treatment duration in days based on
// the method/type string. Conservative defaults so we remind rather than miss.
func estimateTreatmentDurationDays(method string) int {
	lower := strings.ToLower(method)
	has := func(s string) bool { return strings.Contains(lower, s) }
	switch {
	case has("oxalic") && has("vapor"):
		return 7
	case has("oxalic"):
		return 1 // oxalic acid dribble — single application
	case has("apivar"), has("amitraz"):
		return 42
	case has("apiguard"), has("thymol"):
		return 28
	case has("formic"), has("mite away"):
		return 14
	case has("hopguard"):
		return 30
	case has("checkmite"), has("coumaphos"):
		return 42
	default:
		return 14
	}
}

// ---------------------------------------------------------------------------
// Rule: inspection_due
// ---------------------------------------------------------------------------

func checkInspectionDue(ctx context.Context, pool *pgxpool.Pool, now time.Time) ([]Result, error) {
	const defaultIntervalDays = 14

	rows, err := pool.Query(ctx, `
		SELECT h.id, h.position_label,
		       (SELECT max(i.date) FROM inspections i WHERE i.hive_id = h.id)
		FROM hives h
		WHERE h.status = 'active' AND h.is_archived = false`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []Result
	for rows.Next() {
		var (
			hiveID   string
			label    string
			lastDate *time.Time
		)
		if err := rows.Scan(&hiveID, &label, &lastDate); err != nil {
			return nil, err
		}

		if lastDate == nil {
			// Never inspected
			id := hiveID
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Hive "%s" has never been inspected. Schedule an inspection soon.`, label),
				Priority: "high",
			})
			continue
		}

		overdueDays := daysBetween(*lastDate, now) - defaultIntervalDays
		id := hiveID
		switch {
		case overdueDays > 14:
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Hive "%s" is %d days since last inspection. Immediate inspection recommended.`, label, overdueDays+defaultIntervalDays),
				Priority: "urgent",
			})
		case overdueDays > 7:
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Hive "%s" is %d days since last inspection.`, label, overdueDays+defaultIntervalDays),
				Priority: "high",
			})
		case overdueDays > 0:
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Hive "%s" inspection is overdue by %d days.`, label, overdueDays),
				Priority: "normal",
			})
		}
	}
	return results, rows.Err()
}

// ---------------------------------------------------------------------------
// Rule: treatment_reminder
// ---------------------------------------------------------------------------

// treatmentMethod extracts a display/matching string from one treatments entry.
// Entries may be plain strings or objects with method/name/type/product keys.
func treatmentMethod(t any) string {
	switch v := t.(type) {
	case string:
		return v
	case map[string]any:
		for _, key := range []string{"method", "name", "type", "product"} {
			if s, ok := v[key].(string); ok {
				return s
			}
		}
		if b, err := json.Marshal(v); err == nil {
			return string(b)
		}
		return fmt.Sprint(v)
	default:
		return fmt.Sprint(t)
	}
}

func checkTreatmentReminder(ctx context.Context, pool *pgxpool.Pool, now time.Time) ([]Result, error) {
	// Latest inspection with treatments recorded, per active hive.
	rows, err := pool.Query(ctx, `
		SELECT DISTINCT ON (i.hive_id) i.hive_id, i.date, i.treatments, h.position_label
		FROM inspections i
		JOIN hives h ON h.id = i.hive_id
		WHERE h.status = 'active' AND h.is_archived = false
		  AND i.treatments IS NOT NULL AND jsonb_typeof(i.treatments) <> 'null'
		ORDER BY i.hive_id, i.date DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []Result
	for rows.Next() {
		var (
			hiveID     string
			date       time.Time
			treatments []byte
			label      string
		)
		if err := rows.Scan(&hiveID, &date, &treatments, &label); err != nil {
			return nil, err
		}

		var parsed any
		if err := json.Unmarshal(treatments, &parsed); err != nil {
			continue
		}
		list, ok := parsed.([]any)
		if !ok {
			list = []any{parsed}
		}

		for _, t := range list {
			method := treatmentMethod(t)
			durationDays := estimateTreatmentDurationDays(method)
			daysSinceTreatment := daysBetween(date, now)

			if daysSinceTreatment >= durationDays {
				overdue := daysSinceTreatment - durationDays
				priority := "normal"
				if overdue > 7 {
					priority = "high"
				}
				id := hiveID
				results = append(results, Result{
					HiveID:   &id,
					Message:  fmt.Sprintf(`Treatment "%s" on hive "%s" may need follow-up. Applied %d days ago (estimated %d-day duration).`, method, label, daysSinceTreatment, durationDays),
					Priority: priority,
				})
			}
		}
	}
	return results, rows.Err()
}

// ---------------------------------------------------------------------------
// Rule: equipment_needed
// ---------------------------------------------------------------------------

func checkEquipmentNeeded(ctx context.Context, pool *pgxpool.Pool, _ time.Time) ([]Result, error) {
	// Frame shortage per hive from active deployments: box deployments provide
	// capacity (frames_per_box), frame deployments fill it.
	rows, err := pool.Query(ctx, `
		SELECT d.hive_id, h.position_label, d.quantity - d.quantity_returned,
			t.category, t.frames_per_box
		FROM equipment_deployments d
		JOIN equipment_stock s ON s.id = d.stock_id
		JOIN equipment_types t ON t.id = s.type_id
		JOIN hives h ON h.id = d.hive_id
		WHERE d.date_removed IS NULL AND d.quantity > d.quantity_returned
			AND h.status = 'active' AND h.is_archived = false`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	type shortage struct {
		hiveName       string
		totalCapacity  int
		totalInstalled int
	}
	entries := map[string]*shortage{}
	var order []string // stable iteration order

	for rows.Next() {
		var (
			hiveID       string
			label        string
			quantity     int
			category     string
			framesPerBox *int
		)
		if err := rows.Scan(&hiveID, &label, &quantity, &category, &framesPerBox); err != nil {
			return nil, err
		}
		entry, ok := entries[hiveID]
		if !ok {
			entry = &shortage{hiveName: label}
			entries[hiveID] = entry
			order = append(order, hiveID)
		}
		if category == "box" && framesPerBox != nil && *framesPerBox != 0 {
			entry.totalCapacity += quantity * *framesPerBox
		} else if category == "frame" {
			entry.totalInstalled += quantity
		}
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	var results []Result
	for _, hiveID := range order {
		entry := entries[hiveID]
		short := entry.totalCapacity - entry.totalInstalled
		if short > 0 {
			pct := int(math.Round(float64(short) / float64(entry.totalCapacity) * 100))
			priority := "normal"
			if pct > 50 {
				priority = "high"
			}
			id := hiveID
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Hive "%s" needs %d more frames (%d%% capacity unfilled).`, entry.hiveName, short, pct),
				Priority: priority,
			})
		}
	}
	return results, nil
}

// ---------------------------------------------------------------------------
// Rule: feeder_check
// ---------------------------------------------------------------------------

func checkFeederCheck(ctx context.Context, pool *pgxpool.Pool, now time.Time) ([]Result, error) {
	const defaultCheckDays = 7

	// Active feeders are the explicitly open ones (migration
	// 00007_feeding_lifecycle). Legacy records with no recorded end are
	// 'unverified' and are handled by the dashboard feeding-status row, which
	// carries the evidence needed to close them, instead of generating the
	// same generic feeder reminder forever.
	rows, err := pool.Query(ctx, `
		SELECT f.hive_id, f.date_fed, f.type, h.position_label
		FROM feedings f
		JOIN hives h ON h.id = f.hive_id
		WHERE f.status = 'open'`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []Result
	for rows.Next() {
		var (
			hiveID   string
			dateFed  time.Time
			feedType string
			label    string
		)
		if err := rows.Scan(&hiveID, &dateFed, &feedType, &label); err != nil {
			return nil, err
		}

		daysSinceFed := daysBetween(dateFed, now)
		if daysSinceFed >= defaultCheckDays {
			priority := "normal"
			if daysSinceFed > 14 {
				priority = "high"
			}
			id := hiveID
			results = append(results, Result{
				HiveID:   &id,
				Message:  fmt.Sprintf(`Feeder on hive "%s" (%s) placed %d days ago. Check if empty or needs refill.`, label, strings.ReplaceAll(feedType, "_", " "), daysSinceFed),
				Priority: priority,
			})
		}
	}
	return results, rows.Err()
}

// ---------------------------------------------------------------------------
// Rule: seasonal_prep
// ---------------------------------------------------------------------------

func checkSeasonalPrep(ctx context.Context, pool *pgxpool.Pool, now time.Time) ([]Result, error) {
	var (
		seasonalMessage string
		priority        string
	)
	switch now.Month() {
	case time.February, time.March: // Spring buildup
		seasonalMessage = "Spring buildup season. Check food stores, assess colony strength, and consider supplemental feeding if stores are low."
		priority = "normal"
	case time.April, time.May: // Swarm season
		seasonalMessage = "Swarm season alert. Inspect for swarm cells, ensure adequate space, and consider splits for strong colonies."
		priority = "high"
	case time.September, time.October: // Fall prep
		seasonalMessage = "Fall preparation time. Assess winter stores, treat for varroa if needed, and reduce entrances."
		priority = "high"
	case time.December, time.January: // Winter check
		seasonalMessage = "Winter monitoring period. Check hive weight for adequate stores, ensure ventilation, and minimize disturbance."
		priority = "low"
	default:
		return nil, nil
	}

	rows, err := pool.Query(ctx, `
		SELECT h.id, h.position_label
		FROM hives h
		WHERE h.status = 'active' AND h.is_archived = false`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []Result
	for rows.Next() {
		var hiveID, label string
		if err := rows.Scan(&hiveID, &label); err != nil {
			return nil, err
		}
		id := hiveID
		results = append(results, Result{
			HiveID:   &id,
			Message:  fmt.Sprintf(`%s (Hive: %s)`, seasonalMessage, label),
			Priority: priority,
		})
	}
	return results, rows.Err()
}
